Libraries/LLM/Tokenizer.swift

```diff
@@ -41,3 +41,49 @@ let replacementTokenizers = [
     "Qwen2Tokenizer": "PreTrainedTokenizer",
     "CohereTokenizer": "PreTrainedTokenizer",
 ]
+
+public protocol StreamingDetokenizer: IteratorProtocol<String> {
+
+    mutating func append(token: Int)
+
+}
+
+public struct NaiveStreamingDetokenizer: StreamingDetokenizer {
+    let tokenizer: Tokenizer
+
+    var segmentTokens = [Int]()
+    var segment = ""
+
+    public init(tokenizer: Tokenizer) {
+        self.tokenizer = tokenizer
+    }
+
+    mutating public func append(token: Int) {
+        segmentTokens.append(token)
+    }
+
+    mutating func startNewSegment() {
+        let lastToken = segmentTokens.last
+        segmentTokens.removeAll()
+        if let lastToken {
+            segmentTokens.append(lastToken)
+            segment = tokenizer.decode(tokens: segmentTokens)
+        } else {
+            segment = ""
+        }
+    }
+
+    public mutating func next() -> String? {
+        let newSegment = tokenizer.decode(tokens: segmentTokens)
+        let new = newSegment.suffix(newSegment.count - segment.count)
+
+        if new.contains("\n") {
```

Comment:
Shouldn't that be ends with?

Reply:
No, and perhaps that is a bug with the python detokenizer.  Adding a token will go from `foo` to `foo\n\nbar`

Reply:
Hmm... but in that case it might not be safe to split the stream after `\n\nbar`? The implicit assumption is that whenever the text ends in `\n` the detokenization is always the same whether it is done on the full string or split into two chunks at `\n`.

Reply:
Let me get an example -- maybe we need to discuss this.

Reply:
For the first paragraph in the example it produces this:

tokens: [1954, 22094, 29892, 9425, 1006, 2029, 3406, 29892, 393, 12595, 14973, 1063, 8608, 287, 1549, 278, 286, 2879, 310, 931, 304, 385, 21502, 305, 988, 521, 2561, 719, 20913, 29879, 322, 3514, 332, 1794, 696, 314, 278, 12625, 310, 1238, 566, 284, 5546, 29889, 2567, 29892, 7623, 393, 266, 457, 10435, 322, 5075, 272, 505, 1063, 8611, 411, 263, 16624, 936, 25792, 29892, 20794, 14904, 304, 263, 1855, 29885, 310, 1095, 2222, 24496, 322, 7134, 29889, 512, 445, 1855, 29885, 29892, 2012, 310, 8121, 303, 292, 363, 10657, 322, 20123, 29892, 13874, 21126, 338, 304, 16508, 29714, 322, 3957, 4822, 21188, 12625, 29889, 29871, 13, 13]

string: "Imagine, dear interlocutor, that thou hast been transported through the mists of time to an epoch where chivalry reigns and samurai roam the lands of feudal Japan. Now, picture that thine horse and armor have been replaced with a mystical portal, bringing thee to a realm of endless possibilities and knowledge. In this realm, instead of jousting for honor and territory, thy quest is to seek wisdom and connection across distant lands. "

The next token is `\n\nThis`:

tokens: [1954, 22094, 29892, 9425, 1006, 2029, 3406, 29892, 393, 12595, 14973, 1063, 8608, 287, 1549, 278, 286, 2879, 310, 931, 304, 385, 21502, 305, 988, 521, 2561, 719, 20913, 29879, 322, 3514, 332, 1794, 696, 314, 278, 12625, 310, 1238, 566, 284, 5546, 29889, 2567, 29892, 7623, 393, 266, 457, 10435, 322, 5075, 272, 505, 1063, 8611, 411, 263, 16624, 936, 25792, 29892, 20794, 14904, 304, 263, 1855, 29885, 310, 1095, 2222, 24496, 322, 7134, 29889, 512, 445, 1855, 29885, 29892, 2012, 310, 8121, 303, 292, 363, 10657, 322, 20123, 29892, 13874, 21126, 338, 304, 16508, 29714, 322, 3957, 4822, 21188, 12625, 29889, 29871, 13, 13, 4013]

string: "Imagine, dear interlocutor, that thou hast been transported through the mists of time to an epoch where chivalry reigns and samurai roam the lands of feudal Japan. Now, picture that thine horse and armor have been replaced with a mystical portal, bringing thee to a realm of endless possibilities and knowledge. In this realm, instead of jousting for honor and territory, thy quest is to seek wisdom and connection across distant lands. 

This"

Reply:
The next paragraph starts like this:

tokens: [4013, 1855]
string: "This real"

tokens: [4013, 1855, 29885]
string: "This realm"

So it is the sequence of tokens that produces the `\n\n`, not the 4013 itself.

Reply:
Offline discussion:
- python is seeing newlines at the end of the string (I thought it wasn't but I can't be sure what I was looking at now)
- swift (different tokenizer implementation) sees newlines at the end with some models and not others
- we will use `hasSuffix()` to match the python behavior